### ISSUE TITLE
Slow Transfer Speed

### ISSUE THREAD

**Issue by Ohmann-tech** [warner/magic-wormhole#381]
The sender and the receiver are both installed on Windows. The average Speed on both side is about 450kB/s and is pretty slow... The Upload maximum of the sender side is 50mbit/s and the wormhole speed doesnt depend on the number of wormholes. With two wormholes i still have the same speed. 
Both our machines are capable of a lot more.
![image](https://user-images.githubusercontent.com/56092003/78359256-fc8ec700-75b4-11ea-8d75-a0bdf32bcdc9.png)

**Comment by warner** [comment]
Huh, that's weird. I can get at least tens of MBps (note bytes) with a direct connection.

I see that your transfer was going through the Transit Relay, which happens when neither side has a public IP address (and because we haven't yet implemented any form of NAT-hole-punching (#6)). The transit relay is pretty beefy, and I see folks getting 20-50Mbps (note bits) on a regular basis, so maybe 2-7MBps. But the relay is a shared resource, everybody in the world transferring data who can't use a direct connection will go through the same relay, so it's conceivable that you collided with multiple other users pushing a lot of data at the same time, which would lower your speed.

The other thing to look at is the CPU usage of the `wormhole` process. I wouldn't expect it to be noticeable until you were transferring multiple MBps (especially if that box is capable of sending 50mbps without wormhole). But if it's very high, then maybe the transmit process is super-inefficient on your machine and you're now CPU-limited somehow.

If it's the transit relay being a bottleneck, the most immediate fix would be to try to transfer to or from a machine with a public IP address, which would enable a direct connection, bypassing the transit relay. The better fix is to implement hole-punching (#6), but I don't know exactly how to do that yet.

**Comment by nrdxp** [comment]
Just experienced this today. I have a 40Mb upload but the transfer rate of a recent send was only about 30Kb. We reverted to sending the file via our internal slack and it took me less than a minute to upload and the user downloaded in 20 seconds.

**Comment by KyodaiKen** [comment]
My IP is a public IP but it still doesn't work. Do I have to open a port? If yes, which one?

**Comment by meejah** [comment]
@KyodaiKen you got a Transit connection despite a public IP, you mean? When you say you have a public IP, is the very computer you're transferring from on such an address? (Also, though, even if you do have a transit connection it should still be fairly snappy as per prior comment).

**Comment by KyodaiKen** [comment]
Yes, I got Dual-Stack Pro, with both a dedicated (but dynamic) IPv4 and a dedicated (also dynamic) IPv6 and the PC is connected via a router modem combination for cable TV internet (DOCSIS 3.0 and 3.1 hybrid) and my PC is connected to that router directly via Gigabit Ethernet. And we got the very same slow speed through the relay server here. It's Europe, Germany <> Luxemburg.

**Comment by meejah** [comment]
That is likely a NAT setup to your actual PC, then?

The comments about about "a public IP address" mean that the very computer running `wormhole` is directly on a public IP address.

The relay is in North America I believe. It _should_ be possible to use the LeastAuthority / Winden relay (which is in Germany) if you want to try that? See https://winden.app/faq for the address. You should be able to use just the relay (i.e. continue using the default mailbox server).

**Comment by KyodaiKen** [comment]
Seems like for most people they are unlucky with magic wormhole to get P2P at the current setup. Because I think like 99.9% of people are behind a router, not directly connected to the internet unless they use their NAS as a router and install magic wormhole on it.

I can try the leastauthority relay server tomorrow.

**Comment by piegamesde** [comment]
Maybe also try the Rust implementation, as it does quite a bit of effort to work around NATs and firewalls.

**Comment by meejah** [comment]
@KyodaiKen yes, many computers are still behind NATs unfortunately. Still, you should be able to get decent throughput via the relay. If you have concrete numbers for your trials, please post them :)
